docs/ref/contrib/postgres/constraints.txt

```diff
@@ -0,0 +1,48 @@
+========================================
+PostgreSQL specific database constraints
+========================================
+
+.. module:: django.contrib.postgres.constraints
+   :synopsis: PostgreSQL specific database constraint
+
+The exclusion constraint is like :class:`~django.db.models.CheckConstraint`,
+it is useful to ensure different integrity constraints on :ref:`range fields
+<range-fields>`, or tables that have implemented a range with two fields. They
+are declared in :attr:`Meta.constraints
+<django.db.models.Options.constraints>`, and detected automatically when you
+run :djadmin:`makemigrations`. The exclusion constraint may be considered as an
+advanced unique constraint.
+
+You need to activate the `btree_gist extension
+<https://www.postgresql.org/docs/current/static/btree-gist.html>`_ on
+PostgreSQL before this constraint can be used as described in the
+documentation_. You can install it using the
+:class:`~django.contrib.postgres.operations.BtreeGistExtension` migration
+operation.
+
+The classic example discussed in the documentation_ uses a hotel and excludes
+overlapping reservations with check constraints. The constraint also supports
+a condition to limit what is checked.
+
+.. _documentation: https://www.postgresql.org/docs/current/static/sql-createtable.html#SQL-CREATETABLE-EXCLUDE
+
+``ExclusionConstraint``
+=======================
+
+.. class:: ExclusionConstraint(*, expressions=(), index_type='GIST', condition=None, name='')
+
+    The exclusion constraint is implemented using an index. By default, the
+    exclusion constraint is constructed using
+    :class:`~django.contrib.postgres.indexes.GistIndex`. The ``expressions``
+    argument takes a iterable of tuples, where the first element is an
+    expression, string or ``F``-object, and the other element is an operator
+    represented as a string. The ``condition`` argument takes a
+    :class:`~django.db.models.Q`-instance as input. The argument ``name`` is
+    required.
+
+    If the table you want to add an exclusion constraint for has implemented a
+    range using two fields, instead of a :ref:`range field <range-fields>`, you
+    should write an expression that uses the equivalent function, and use the
+    delimiters for the field. Most often, the delimiters will be ``[)``,
+    meaning that the lower bound will be included, and the upper bound
+    excluded.
```

Comment:
A code example of this would probably be useful.

Reply:
Agreed.

Reply:
There's also a request (and a possible solution in that ticket) for adding support for defining custom-bounds on the range-types: https://code.djangoproject.com/ticket/27147